### DISCUSSION TITLE
Allow exceptions and logging in render function

### DISCUSSION THREAD

**Discussion by whittle** [srid/ema#63] [Ideas]
The `runEma` and `runEmaWithCli` functions expect the `Asset` renderer passed to be a pure function. It would be helpful to be able to raise exceptions or log messages from the render action.

**Comment by srid** [comment]
@whittle What sort of errors would your render function be raising? Just trying to better understand the original goal of this solution.

**Reply by whittle** [reply]
Currently, I’m working on a static wiki-alike generator, and I started from srid/ema-template. srid/ema-template uses `Control.Exception.throw` to raise exceptions from the `rpBlock` and `rpInline` functions, which are part of the render pathway. In my generator, I’d like to avoid throwing IO exceptions from pure code, even if it does muddy the type signatures. 

When I originally made this request, I also wanted to be able to raise exceptions from the Pandoc transformations that happen in the render pathway, but I’m currently working on moving those to the modeling thread.

**Reply by srid** [reply]
>  I’m currently working on moving those to the modeling thread.

Right ... generally speaking, it is best to do all the impure stuff inside the `runEma` block where the model is being (incrementally) built. And when it comes to the pure `render` function it is expected to terminate successfully with the asset value (unless when raising errors that are not to be handled).

But I can see the value of having Ema take a monadic render function, with most users just doing `Ema.runEma (runIdentity . render)` or something like that.

**Comment by srid** [comment]
The render function (`siteOutput)` in master branch supports IO, so this should no longer be an issue, @whittle 

I'm yet to release 0.8, but docs are more or less up to date: https://ema.srid.ca/
